main.go

```diff
@@ -17,121 +17,163 @@ package main
-		os.Exit(1)
+func runOperator(ctx context.Context, controllerContext *controllercmd.ControllerContext) error {
+	ns := os.Getenv("COMPONENT_NAMESPACE")
+	if ns == "" {
+		if controllerContext.ComponentConfig != nil {
+			ns = controllerContext.ComponentConfig.GetNamespace()
+		}
+		if ns == "" {
+			ns = controllers.ComponentNamespace
+		}
 	}
 
-	if err := osconfigv1.AddToScheme(scheme); err != nil {
-		klog.ErrorS(err, "Error adding k8s client to scheme.")
-		os.Exit(1)
+	client, err := metal3ioClient.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
 	}
+	metal3Informers := metal3externalinformers.NewSharedInformerFactory(client, 10*time.Minute)
 
-	// +kubebuilder:scaffold:scheme
-	// The following is needed to read the Infrastructure CR
-	if err := osconfigv1.Install(scheme); err != nil {
-		klog.ErrorS(err, "")
-		os.Exit(1)
+	kubeClient, err := kubernetes.NewForConfig(controllerContext.KubeConfig)
+	if err != nil {
+		return err
 	}
-}
-
-func main() {
-	var metricsAddr string
-	var enableLeaderElection bool
-	var imagesJSONFilename string
-
-	klog.InitFlags(nil)
-	flag.StringVar(&metricsAddr, "metrics-addr", ":8080", "The address the metric endpoint binds to.")
-	flag.BoolVar(&enableLeaderElection, "enable-leader-election", false,
-		"Enable leader election for controller manager. Enabling this will ensure there is only one active controller manager.")
-	flag.StringVar(&imagesJSONFilename, "images-json", "/etc/cluster-baremetal-operator/images/images.json",
-		"The location of the file containing the images to use for our operands.")
-	flag.Parse()
+	kubeInformersForNamespace := informers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, informers.WithNamespace(ns))
 
-	ctrl.SetLogger(klogr.New())
+	osClient, err := osclientset.NewForConfig(rest.AddUserAgent(controllerContext.KubeConfig, "config-shared-informer"))
+	if err != nil {
+		return err
+	}
+	configInformers := configv1informers.NewSharedInformerFactory(osClient, 10*time.Minute)
 
 	releaseVersion := os.Getenv("RELEASE_VERSION")
 	if releaseVersion == "" {
 		klog.Info("Environment variable RELEASE_VERSION not provided")
 	}
 
-	config := ctrl.GetConfigOrDie()
-	mgr, err := ctrl.NewManager(config, ctrl.Options{
-		Scheme:             scheme,
-		MetricsBindAddress: metricsAddr,
-		LeaderElection:     enableLeaderElection,
-		Port:               9443,
-		CertDir:            "/etc/cluster-baremetal-operator/tls",
-	})
-	if err != nil {
-		klog.ErrorS(err, "unable to start manager")
-		os.Exit(1)
-	}
+	provisioningController := controllers.NewProvisioningController(
+		client,
+		kubeClient,
+		osClient,
+		kubeInformersForNamespace,
+		metal3Informers,
+		configInformers,
+		controllerContext.EventRecorder,
+		releaseVersion,
+		imagesJSONFilename,
+	)
+
+	metal3Informers.Start(ctx.Done())
+	kubeInformersForNamespace.Start(ctx.Done())
+	configInformers.Start(ctx.Done())
+
+	go provisioningController.Run(ctx, 1)
```

Comment:
To aggregate status from FooDegraded, FooAvailable on a resource like kubeapiserver.operator.openshift.io to clusteroperator.config.openshift.io/kube-apiserver we use https://github.com/openshift/library-go/blob/7a65fdb398e28782ee1650959a5e0419121e97ae/pkg/operator/status/status_controller.go#L59-L67 .

This allows us to add various degraded checks easily.

Reply:
@deads2k that controller is not very obvious.. still trying to figure out what exactly it does

Reply:
> > > There is a real e2e failure that amounts to
> > > ```
> > > oc explain provisionings --api-version=metal3.io/v1alpha1
> > > error: couldn't find resource for "metal3.io/v1alpha1, Kind=AdmissionReview"
> > > ```
> > > 
> > > 
> > > also...
> > > ```
> > > oc explain provisionings
> > > error: couldn't find resource for "admission.metal3.io/v1alpha1, Kind=AdmissionReview"
> > > ```
> > 
> > 
> > You'll want `oc get --raw /apis/admission.metal3.io/v1alpha1` and have @soltysh check the content to see if it's an `oc explain` bug (very possible) or a discovery bug.
> 
> oc get --raw /apis/admission.metal3.io/v1alpha1
> {"kind":"APIResourceList","apiVersion":"v1","groupVersion":"admission.metal3.io/v1alpha1","resources":[{"name":"provisionings","singularName":"","namespaced":false,"group":"admission.k8s.io","version":"v1beta1","kind":"AdmissionReview","verbs":["create"]}]}

I have worked around this by changing the APIService/Webhook path (I had a look at what hive was doing) https://github.com/openshift/cluster-baremetal-operator/pull/129/commits/2d1f5ef5d1b0f178b50ef12fe7677c915dd07236